### DISCUSSION TITLE
OpenSSL in FIPS mode in statically linked applications doesn't work on FreeBSD?

### DISCUSSION THREAD

**Discussion by StephenWall** [openssl/openssl#23830] [Q&A] [answered]
FreeBSD 14.0p5 with provided OpenSSL 3.0.12, FIPS provider both 3.0.12 and 3.0.9

While investigating a problem I encountered using FreeBSD's `pkg-static` command (`pkg` is the FreeBSD package management application, `pkg-static` is a statically linked version of it, so that packages can be managed even if your system gets out of whack), I learned from some FreeBSD folks that `dlopen()`, the function used by OpenSSL to load the FIPS provider on FreeBSD, does not function in statically linked applications.

This means that FIPS mode cannot work with static linking on FreeBSD (I can't speak to other platforms).

Is this a known limitation?  I don't recall seeing anything like this when using OpenSSL 1.0.2 / FIPS engine 2.x on older versions of FreeBSD.

**Comment by t8m** [comment, marked as answer]
If FreeBSD does not support loading dynamic modules in statically linked applications it cannot support 3.x FIPS modules in such scenario as the FIPS module has to be loaded dynamically. There is no way around that.

**Comment by paulidale** [comment]
OpenSSL with FIPS as of version 3.0 is not possible when statically linked.
Live with it.  The static linking option using 1.0.2 was a gross hack which is no longer possible to support.
Honestly, plenty of platforms are no longer able to support the FIPS provide.

If you want static linking or other platform support using the FIPS provider, pease submit suitable patches for consideration.
The project will absolutely consider any such submission.
